src/npz.rs

```diff
@@ -243,3 +253,545 @@ impl<R: Read + Seek> NpzReader<R> {
+                    },
+                );
+            } else {
+                return Err(ZipError::InvalidArchive("Ambiguous offsets").into());
+            }
+        }
+        Ok(Self { files, names })
+    }
+
+    /// Returns `true` iff the `.npz` file doesn't contain any **uncompressed** arrays.
+    pub fn is_empty(&self) -> bool {
+        self.names.is_empty()
+    }
+
+    /// Returns the number of **uncompressed** arrays in the `.npz` file.
+    pub fn len(&self) -> usize {
+        self.names.len()
+    }
+
+    /// Returns the names of all of the **uncompressed** arrays in the file.
+    pub fn names(&self) -> impl Iterator<Item = &str> {
+        self.names.keys().map(String::as_str)
+    }
+
+    /// Moves a mutable `.npy` file view by name out of the `.npz` file view.
+    pub fn by_name(&mut self, name: &str) -> Result<NpyViewMut<'a>, ViewNpzError> {
+        self.by_index(
+            self.names
+                .get(name)
+                .copied()
+                .ok_or(ZipError::FileNotFound)?,
+        )
+    }
+
+    /// Moves a mutable `.npy` file view by index in `0..len()` out of the `.npz` file view.
+    pub fn by_index(&mut self, index: usize) -> Result<NpyViewMut<'a>, ViewNpzError> {
+        if index > self.names.len() {
+            Err(ZipError::FileNotFound.into())
+        } else {
+            self.files
+                .remove(&index)
+                .ok_or(ViewNpzError::MovedNpyViewMut)
+        }
+    }
+}
+
+/// Mutable view of memory-mapped `.npy` files within an `.npz` file.
+///
+/// Does **not** automatically [verify](`Self::verify`) the CRC-32 checksum but **does**
+/// [update](`Self::update`) it on [`Drop::drop`].
+#[derive(Debug)]
+pub struct NpyViewMut<'a> {
+    crc32: &'a mut [u8],
+    data: &'a mut [u8],
+    central_crc32: &'a mut [u8],
+}
+
+impl<'a> NpyViewMut<'a> {
+    /// Verifies CRC-32 checksum by reading the whole array.
+    pub fn verify(&self) -> Result<(), ViewNpzError> {
+        // Like the `zip` crate, verify only against central CRC-32.
+        Ok(crc32_verify(self.data, self.central_crc32)?)
+    }
+
+    /// Returns an immutable view of a memory-mapped `.npy` file.
+    ///
+    /// Iterates over `bool` array to ensure `0x00`/`0x01` values.
+    pub fn view<A, D>(&self) -> Result<ArrayView<A, D>, ViewNpzError>
+    where
+        A: ViewElement,
+        D: Dimension,
+    {
+        Ok(ArrayView::<A, D>::view_npy(self.data)?)
+    }
+
+    /// Returns a mutable view of a memory-mapped `.npy` file.
+    ///
+    /// Iterates over `bool` array to ensure `0x00`/`0x01` values.
+    pub fn view_mut<A, D>(&mut self) -> Result<ArrayViewMut<A, D>, ViewNpzError>
+    where
+        A: ViewMutElement,
+        D: Dimension,
+    {
+        Ok(ArrayViewMut::<A, D>::view_mut_npy(self.data)?)
+    }
+
+    /// Updates CRC-32 checksum by reading the whole array.
+    ///
+    /// Automatically updated on [`Drop::drop`].
+    pub fn update(&mut self) {
+        self.central_crc32
```

Comment:
It may be worth tracking whether the data has changed since the last update. There are three cases of interest.

```rust
enum ChecksumStatus {
    /// The checksum has not been computed, and the data has not changed.
    Unverified,
    /// The checksum is correct, and the data has not changed.
    Correct,
    /// The data may have changed.
    Outdated,
}
```

1. The data is unchanged, and the checksum has never been computed by `NpyViewMut`. In this case, a few of the options we can choose are the following:

   1. Assume that the original checksum in the file is correct and skip recomputing it.
   2. Compute the checksum and write it to the file. Don't check whether it matches the original checksum.
   3. Compute the checksum. If it does not match the original checksum, write it to the file and return an error.
   4. Compute the checksum. If it does not match the original checksum, return an error without writing it to the file.

   I prefer option iii, but I'd be interested in your thoughts. The user can ignore the error if they want. We should mention in the docs that the `Drop` impl ignores the error.

2. The data is unchanged, and the checksum has been verified as correct. In this case, we can skip recomputing the checksum.

3. The data may have changed. In this case, we should compute the checksum and write it to the file.

The flag would be initialized to `Unverified`. `verify` would change the flag to `Correct` in the `Ok` case. `view_mut` would change the flag to `Outdated`. `update` would change the flag to `Correct`.

Reply:
I found it easier to document by making `ChecksumStatus` public and then did it a bit differently in the end. I think it's practical when `Drop` only reads the array if the checksum is `Outdated`. Most use cases have correct checksums and don't verify or update it manually as modern file systems do this already on their own. The methods are explicit this way and always do what they are named like. What do you think?